### DISCUSSION TITLE
Usage of bevy game engine

### DISCUSSION THREAD

**Discussion by maxammann** [maplibre/maplibre-rs#75] [General]
We have to think about the rendering architecture soon. The [bevy](https://bevyengine.org/) engine brings already such a structure.
I think we will need some kind of render graph like implemented here by bevy: https://github.com/bevyengine/bevy/tree/main/crates/bevy_render/src/render_graph

I'm a little bit undecided whether using bevy is a good idea. Let's collect here some pro/contras of using bevy.

## Pro

* Reusage of bevy plugins
* Reusage of rendering infrastructure

## Contra

* Hard dependency on winit
* No direct access to low-level wgpu API 
* Possibly more difficult to port to iOS/Android/web in the way we want
* Harder to optimize

**Comment by maxammann** [comment]
## Contra

* Most features of bevy are game related and not relevant for us
  * We probably wont use features like: PBR, advanced shading, lots of postprocessing stages, ECS, etc 
  * We would only use the render graph + sprite management from bevy
 * When hitting limitations with bevy, we are in direct concurrence with other developers. There could be a situation in which we need a bevy feature in a certain way. Maybe that would be uninteresting for game development though. That means i probably won't get implemented.

**Reply by DerKarlos** [reply]
I assume one could bypass bevy and use wgpu directly to.
And you may use PBR etc. if one fare day, maplibre-rs will support 3D and VR
(I just transcoded some of the OSM2Word Web-Frontend to RUST. More soon)

**Reply by maxammann** [reply]
Bevy does not support this unfortunately. Basically it was already decided not to use bevy. Bevy has soo so many cool features which are not relevant for maplibre-rs. So I think it makes sense to get inspiration from bevy, but not using it directly.

This looks indeed very cool: http://osm2world.org/

What exactly did you do there in Rust?

**Reply by DerKarlos** [reply]
Ah, I transcoded parts of the OSM2World-renderer to Rust, used rend3 OR bevy to get it on the screen and to move around.
You may see it here: https://github.com/DerKarlos/O2W-Rust-WASM or even build and run it.
It was tricky to get it independent of the rendere engines. I wonder, if I could do this for mablibre-rs (_MRS_) to.

The Bevy renderer was updated recently. May be _MRS_ should update too.
Bevy was split in modules, so you don't need to use it all. May be it could be used now for _MRS_? (I need to look at the concepts of _MRS_!)

**Reply by maxammann** [reply]
The repo seems to be private

**Reply by DerKarlos** [reply]
Yes, because it's not presentable yet. But you could get a collaborator easily.
Time went by, I have to get into my and your project again now ...

**Reply by maxammann** [reply]
Great! I have not yet found time, but let's see!

**Comment by therishidesai** [comment]
It looks like maplibre has made the decision to not use bevy. But with #253 maplibre now has its own ECS like system based off of bevy. Is the goal for maplibre to be its own self contained app building ecosystem strictly for maps? Main reason I ask is I am looking into using Rust + WASM to build video games that can render real world maps. I would like to combine maplibre’s rendering with the rest of the features Bevy has to offer. Is this possible with maplibre today?
